### DISCUSSION TITLE
Ensuring normalisation at all times of multiple components (variables)

### DISCUSSION THREAD

**Discussion by RobMacka** [idaholab/moose#20229] [Q&A General] [answered]
Hi all,

I wish to ensure the quadratic sum of multiple variables (representing vector components) equals 1 at all times during the simulation. I need this normalisation to be compatible with seperate evolution with free energy functionals with the Phase Field module. Thank you for your time and help.

Thanks,
Robert

**Comment by GiudGiud** [comment, marked as answer]
@lindsayad I m thinking he should work with a lagrange multiplier scalar variable to enforce the constraint. Can you think of alternatives?

**Reply by lindsayad** [reply]
I agree

**Reply by RobMacka** [reply]
Thank you both for your responses. But doesn't the ScalarLagrangeConstraint only work with the variables themselves and not functions of them? As I need mine to work with their quadratic sum, thanks again

**Reply by GiudGiud** [reply]
Hello

There may not be an object that does exactly what you want already. You will likely have to build it off of existing objects. 
https://mooseframework.inl.gov/source/kernels/ScalarLagrangeMultiplier.html is almost what you want, you just need to change the implementation of the norm to the one you want. 

Guillaume

**Reply by GiudGiud** [reply]
actually you ll need to adapt it to array variables as well, since it's only looking at a single variable.

**Reply by mangerij** [reply]
What vector does this represent? 

We have done this for magnetization successfully using a longitudinal term in the PDE. Perhaps something similar can work for you. I suppose it depends on what equation governs your vector component variables

Another alternative is letting your vector live on the unit sphere and reformulate your PDE in terms of angles. If the PDE isn't too horrible, it would work and you get the added bonus of having one less variable and significantly less dof when you scale.
